In [1]:
#!pip install skorch

In [2]:
import itertools
import numpy as np
import pandas as pd
#from skorch import NeuralNetClassifier
import torch
import torch.nn as nn
from torch.utils.data import (
    Dataset,
    DataLoader,
    SubsetRandomSampler,
    TensorDataset
    )
from torch.optim import Optimizer
from sklearn.model_selection import (
    train_test_split,
    KFold,
    GridSearchCV,
    RandomizedSearchCV
  )
from sklearn.metrics import(
    accuracy_score,
    precision_recall_fscore_support,
    roc_auc_score
    )
from sklearn.base import BaseEstimator
from sklearn.preprocessing import StandardScaler
from imblearn.over_sampling import SMOTE
from transformers import GPT2LMHeadModel, GPT2Tokenizer, GPT2Config
import xgboost as xgb
from tqdm import tqdm

In [3]:
# Custom Dataset
class MortgageDataset(Dataset):
    def __init__(self, features, labels):
        self.features = torch.tensor(features, dtype=torch.float32)
        self.labels = torch.tensor(labels, dtype=torch.long)

    def __len__(self):
        return len(self.labels)

    def __getitem__(self, idx):
        return self.features[idx], self.labels[idx]


# Custom Optimizer
class AdvancedSymplecticOptimizer(Optimizer):
    def __init__(self, params, lr=1e-2, beta=0.9, epsilon=1e-8):
        defaults = dict(lr=lr, beta=beta, epsilon=epsilon)
        super(AdvancedSymplecticOptimizer, self).__init__(params, defaults)

    @torch.no_grad()
    def step(self, closure=None):
        loss = None
        if closure is not None:
            with torch.enable_grad():
                loss = closure()

        for group in self.param_groups:
            for p in group['params']:
                if p.grad is None:
                    continue
                grad = p.grad
                state = self.state[p]

                if len(state) == 0:
                    state['step'] = 0
                    state['momentum'] = torch.zeros_like(p.data)

                momentum = state['momentum']
                lr, beta, eps = group['lr'], group['beta'], group['epsilon']

                state['step'] += 1

                # Implement 4th order symplectic integrator (Forest-Ruth algorithm)
                momentum.mul_(beta).add_(grad, alpha=1 - beta)

                # Compute adaptive step size
                kinetic = 0.5 * (momentum ** 2).sum()
                potential = 0.5 * (grad ** 2).sum()
                hamiltonian = kinetic + potential
                step_size = lr / (hamiltonian.sqrt() + eps)

                p.add_(momentum, alpha=-step_size)

        return loss


class HamiltonianNN(nn.Module):
    def __init__(self, input_dim, hidden_dims=[512, 268], activation='leaky_relu', dropout_rate=0.2):
        super().__init__()
        self.layers = nn.ModuleList()

        # Input layer
        self.layers.append(nn.Linear(input_dim, hidden_dims[0]))

        # Hidden layers
        for i in range(len(hidden_dims) - 1):
            self.layers.append(nn.Linear(hidden_dims[i], hidden_dims[i+1]))

        # Output layer
        self.layers.append(nn.Linear(hidden_dims[-1], 2))  # Binary classification

        # Activation function
        self.activation = nn.LeakyReLU()
        self.dropout = nn.Dropout(dropout_rate)

    def forward(self, x):
        for layer in self.layers[:-1]:  # All layers except the last
            x = self.dropout(self.activation(layer(x)))
        return self.layers[-1](x)  # Last layer (no activation for output)


# Hamiltonian-inspired loss function
def hamiltonian_loss(outputs, labels, model, reg_coeff=0.01):
    loss_fct = nn.CrossEntropyLoss()
    base_loss = loss_fct(outputs, labels)
    # Add regularization based on Hamiltonian principles
    param_norm = sum(p.norm().item() for p in model.parameters())
    reg_term = reg_coeff * param_norm  # Use reg_coeff instead of a fixed value
    return base_loss + reg_term


# Evaluation function
def evaluate_model(model, dataloader, device):
    model.eval()
    all_preds = []
    all_labels = []

    with torch.no_grad():
        for features, labels in dataloader:
            features, labels = features.to(device), labels.to(device)
            outputs = model(features)
            _, preds = torch.max(outputs, 1)
            all_preds.extend(preds.cpu().numpy())
            all_labels.extend(labels.cpu().numpy())

    accuracy = accuracy_score(all_labels, all_preds)
    precision, recall, f1, _ = precision_recall_fscore_support(all_labels, all_preds, average='weighted', zero_division=0)
    auc = roc_auc_score(all_labels, all_preds)

    return accuracy, precision, recall, f1, auc

In [4]:
# Load the data
train_df = pd.read_csv("Training.csv")
oot_test_df = pd.read_csv("OOT test.csv")
train_df = train_df.dropna()
oot_test_df = oot_test_df.dropna()

# Prepare features and target
features = ['Credit_Score', 'Mortgage_Insurance', 'Number_of_units', 'CLoan_to_value',
            'Debt_to_income', 'OLoan_to_value', 'Single_borrower']

X_train = train_df[features]
y_train = train_df['DFlag']
X_oot_test = oot_test_df[features]
y_oot_test = oot_test_df['DFlag']

# Scale the features
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_oot_test_scaled = scaler.transform(X_oot_test)

# Apply SMOTE to balance the training dataset
smote = SMOTE(random_state=42)
X_train_resampled, y_train_resampled = smote.fit_resample(X_train_scaled, y_train)

# Convert to PyTorch tensors
X_train_tensor = torch.FloatTensor(X_train_resampled)
y_train_tensor = torch.LongTensor(y_train_resampled)
X_oot_test_tensor = torch.FloatTensor(X_oot_test_scaled)
y_oot_test_tensor = torch.LongTensor(y_oot_test.values)

# Create DataLoader for training
train_dataset = TensorDataset(X_train_tensor, y_train_tensor)
train_loader = DataLoader(train_dataset, batch_size=32, shuffle=True)

# Create DataLoader for OOT testing
oot_test_dataset = TensorDataset(X_oot_test_tensor, y_oot_test_tensor)
oot_test_loader = DataLoader(oot_test_dataset, batch_size=32, shuffle=False)

# Print class distribution before and after SMOTE
print("Class distribution before SMOTE:", np.bincount(y_train))
print("Class distribution after SMOTE:", np.bincount(y_train_resampled))

Class distribution before SMOTE: [315269   9183]
Class distribution after SMOTE: [315269 315269]


In [5]:
# Set up device
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

In [6]:
# Custom LR Scheduler
class CustomLRScheduler:
    def __init__(self, optimizer, factor=0.5, patience=3, min_lr=1e-5):
        self.optimizer = optimizer
        self.factor = factor
        self.patience = patience
        self.min_lr = min_lr
        self.best_loss = float('inf')
        self.bad_epochs = 0

    def step(self, val_loss):
        if val_loss < self.best_loss:
            self.best_loss = val_loss
            self.bad_epochs = 0
        else:
            self.bad_epochs += 1

        if self.bad_epochs > self.patience:
            for param_group in self.optimizer.param_groups:
                param_group['lr'] = max(param_group['lr'] * self.factor, self.min_lr)
            self.bad_epochs = 0
            print(f"Learning rate reduced to {param_group['lr']}")

In [7]:
class HamiltonianNNWrapper(BaseEstimator):
    def __init__(self, hidden_dims=[64, 32], activation='relu', dropout_rate=0.2, lr=0.01, reg_coeff=0.01):
        self.hidden_dims = hidden_dims
        self.activation = activation
        self.dropout_rate = dropout_rate
        self.lr = lr
        self.reg_coeff = reg_coeff
        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

    def fit(self, X, y):
        try:
            X = torch.tensor(X, dtype=torch.float32).to(self.device)
            y = torch.tensor(y, dtype=torch.long).to(self.device)

            self.model = HamiltonianNN(
                input_dim=X.shape[1],
                hidden_dims=self.hidden_dims,
                activation=self.activation,
                dropout_rate=self.dropout_rate
            ).to(self.device)

            optimizer = AdvancedSymplecticOptimizer(self.model.parameters(), lr=self.lr)

            dataset = TensorDataset(X, y)
            loader = DataLoader(dataset, batch_size=32, shuffle=True)

            for epoch in range(10):  # You can adjust the number of epochs
                for features, labels in loader:
                    optimizer.zero_grad()
                    outputs = self.model(features)
                    loss = hamiltonian_loss(outputs, labels, self.model, self.reg_coeff)
                    loss.backward()
                    optimizer.step()

            return self
        except Exception as e:
            print(f"Error during fitting: {str(e)}")
            raise

    def predict(self, X):
        X = torch.tensor(X, dtype=torch.float32).to(self.device)
        self.model.eval()
        with torch.no_grad():
            outputs = self.model(X)
            _, predictions = torch.max(outputs, 1)
        return predictions.cpu().numpy()

## NN hyperparameters optimization

In [8]:
"""class HamiltonianNNWrapper(BaseEstimator):
    def __init__(self, hidden_dims=[64, 32], activation='relu', dropout_rate=0.2, lr=1e-3, reg_coeff=0.01):
        self.hidden_dims = hidden_dims
        self.activation = activation
        self.dropout_rate = dropout_rate
        self.lr = lr
        self.reg_coeff = reg_coeff
        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

    def fit(self, X, y):
        X = torch.tensor(X, dtype=torch.float32).to(self.device)
        y = torch.tensor(y, dtype=torch.long).to(self.device)

        self.model = HamiltonianNN(
            input_dim=X.shape[1],
            hidden_dims=self.hidden_dims,
            activation=self.activation,
            dropout_rate=self.dropout_rate
        ).to(self.device)

        optimizer = AdvancedSymplecticOptimizer(self.model.parameters(), lr=self.lr)

        dataset = TensorDataset(X, y)
        loader = DataLoader(dataset, batch_size=32, shuffle=True)

        for epoch in range(10):  # You can adjust the number of epochs
            for features, labels in loader:
                optimizer.zero_grad()
                outputs = self.model(features)
                loss = hamiltonian_loss(outputs, labels, self.model, self.reg_coeff)
                loss.backward()
                optimizer.step()

        return self

    def predict(self, X):
        X = torch.tensor(X, dtype=torch.float32).to(self.device)
        self.model.eval()
        with torch.no_grad():
            outputs = self.model(X)
            _, predictions = torch.max(outputs, 1)
        return predictions.cpu().numpy()

# Hyperparameter search space
param_dist = {
    'hidden_dims': [[64, 32], [128, 64]],
    'activation': ['relu', 'leaky_relu'],
    'dropout_rate': [0.2, 0.4],
    'lr': [1e-3, 1e-2],
    'reg_coeff': [0.001, 0.01]
}

# Randomized search
search = RandomizedSearchCV(
    HamiltonianNNWrapper(),
    param_distributions=param_dist,
    n_iter=10,
    cv=3,
    scoring='f1',
    n_jobs=-1
)

# Perform the search
search.fit(X_train_tensor.numpy(), y_train_tensor.numpy())

# Print best parameters and score
print("Best parameters:", search.best_params_)
print("Best F1 score:", search.best_score_)

# Train final model with best hyperparameters
best_model = search.best_estimator_"""

'class HamiltonianNNWrapper(BaseEstimator):\n    def __init__(self, hidden_dims=[64, 32], activation=\'relu\', dropout_rate=0.2, lr=1e-3, reg_coeff=0.01):\n        self.hidden_dims = hidden_dims\n        self.activation = activation\n        self.dropout_rate = dropout_rate\n        self.lr = lr\n        self.reg_coeff = reg_coeff\n        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")\n\n    def fit(self, X, y):\n        X = torch.tensor(X, dtype=torch.float32).to(self.device)\n        y = torch.tensor(y, dtype=torch.long).to(self.device)\n\n        self.model = HamiltonianNN(\n            input_dim=X.shape[1],\n            hidden_dims=self.hidden_dims,\n            activation=self.activation,\n            dropout_rate=self.dropout_rate\n        ).to(self.device)\n\n        optimizer = AdvancedSymplecticOptimizer(self.model.parameters(), lr=self.lr)\n\n        dataset = TensorDataset(X, y)\n        loader = DataLoader(dataset, batch_size=32, shuffle=True)\n

## Training the mode with the best hyperparameters
Best parameters:

*   reg_coeff = 0.01,
*   lr = 0.01
*   hidden_dims = [128, 64]
*   dropout_rate = 0.2
*   activation = 'leaky_relu'

In [9]:
# Train final model with best hyperparameters
num_epochs = 30
model = HamiltonianNN(
    input_dim=X_train.shape[1],
    hidden_dims=[512, 256], #[128, 64]
    activation='leaky_relu',
    dropout_rate=0.2
).to(device)


# K-fold Cross-validation
k_folds = 3
kf = KFold(n_splits=k_folds, shuffle=True, random_state=42)

cv_accuracies = []
cv_f1_scores = []

for fold, (train_idx, val_idx) in enumerate(kf.split(X_train_tensor)):
    print(f"\nFold {fold+1}/{k_folds}")

    X_train_fold, X_val_fold = X_train_tensor[train_idx], X_train_tensor[val_idx]
    y_train_fold, y_val_fold = y_train_tensor[train_idx], y_train_tensor[val_idx]

    train_dataset = TensorDataset(X_train_fold, y_train_fold)
    val_dataset = TensorDataset(X_val_fold, y_val_fold)

    train_loader = DataLoader(train_dataset, batch_size=32, shuffle=True)
    val_loader = DataLoader(val_dataset, batch_size=32)

    model = HamiltonianNN(input_dim=X_train.shape[1]).to(device)
    optimizer = AdvancedSymplecticOptimizer(model.parameters(), lr=0.01)
    scheduler = CustomLRScheduler(optimizer)

    best_val_loss = float('inf')
    best_model_state = None

    for epoch in range(num_epochs):
        model.train()
        total_loss = 0

        for features, labels in tqdm(train_loader, desc=f"Epoch {epoch+1}/{num_epochs}"):
            optimizer.zero_grad()
            features, labels = features.to(device), labels.to(device)

            outputs = model(features)
            loss = hamiltonian_loss(outputs, labels, model)
            loss.backward()
            optimizer.step()

            total_loss += loss.item()

        avg_train_loss = total_loss / len(train_loader)
        print(f"Epoch {epoch+1}/{num_epochs}, Average Train Loss: {avg_train_loss:.4f}")

        # Validation
        model.eval()
        val_loss = 0
        with torch.no_grad():
            for features, labels in val_loader:
                features, labels = features.to(device), labels.to(device)
                outputs = model(features)
                val_loss += hamiltonian_loss(outputs, labels, model).item()

        val_loss /= len(val_loader)
        val_accuracy, val_precision, val_recall, val_f1, val_auc = evaluate_model(model, val_loader, device)
        print(f"Validation - Loss: {val_loss:.4f}, Accuracy: {val_accuracy:.4f}, F1: {val_f1:.4f}, AUC: {val_auc:.4f}, Precision: {val_precision:.4f},  Recall: {val_recall:.4f}")

        # Update learning rate
        scheduler.step(val_loss)

        # Save best model
        if val_loss < best_val_loss:
            best_val_loss = val_loss
            best_model_state = model.state_dict().copy()
            print("New best model saved")

    # Load best model for final evaluation
    model.load_state_dict(best_model_state)
    val_accuracy, val_precision, val_recall, val_f1, val_auc = evaluate_model(model, val_loader, device)
    cv_accuracies.append(val_accuracy)
    cv_f1_scores.append(val_f1)
    print(f"Final Validation - Accuracy: {val_accuracy:.4f}, F1: {val_f1:.4f}, AUC: {val_auc:.4f}")

# Load best model for final evaluation
model.load_state_dict(best_model_state)
print("Training completed. Best model loaded.")

val_accuracy, val_precision, val_recall, val_f1, val_auc = evaluate_model(model, val_loader, device)
cv_accuracies.append(val_accuracy)
cv_f1_scores.append(val_f1)
print(f"Final Validation - Accuracy: {val_accuracy:.4f}, F1: {val_f1:.4f}, AUC: {val_auc:.4f}")


Fold 1/3


Epoch 1/30: 100%|██████████| 13137/13137 [00:49<00:00, 267.95it/s]


Epoch 1/30, Average Train Loss: 0.8759
Validation - Loss: 0.8736, Accuracy: 0.6905, F1: 0.6888, AUC: 0.6904, Precision: 0.6946,  Recall: 0.6905
New best model saved


Epoch 2/30: 100%|██████████| 13137/13137 [00:47<00:00, 275.40it/s]


Epoch 2/30, Average Train Loss: 0.8764
Validation - Loss: 0.8755, Accuracy: 0.6914, F1: 0.6911, AUC: 0.6913, Precision: 0.6918,  Recall: 0.6914


Epoch 3/30: 100%|██████████| 13137/13137 [00:45<00:00, 287.36it/s]


Epoch 3/30, Average Train Loss: 0.8784
Validation - Loss: 0.8781, Accuracy: 0.6915, F1: 0.6900, AUC: 0.6914, Precision: 0.6952,  Recall: 0.6915


Epoch 4/30: 100%|██████████| 13137/13137 [00:45<00:00, 286.20it/s]


Epoch 4/30, Average Train Loss: 0.8810
Validation - Loss: 0.8845, Accuracy: 0.6927, F1: 0.6925, AUC: 0.6926, Precision: 0.6929,  Recall: 0.6927


Epoch 5/30: 100%|██████████| 13137/13137 [00:44<00:00, 293.30it/s]


Epoch 5/30, Average Train Loss: 0.8868
Validation - Loss: 0.8873, Accuracy: 0.6940, F1: 0.6930, AUC: 0.6939, Precision: 0.6963,  Recall: 0.6940
Learning rate reduced to 0.005


Epoch 6/30: 100%|██████████| 13137/13137 [00:46<00:00, 280.94it/s]


Epoch 6/30, Average Train Loss: 0.8893
Validation - Loss: 0.8897, Accuracy: 0.6936, F1: 0.6925, AUC: 0.6935, Precision: 0.6962,  Recall: 0.6936


Epoch 7/30: 100%|██████████| 13137/13137 [00:45<00:00, 288.69it/s]


Epoch 7/30, Average Train Loss: 0.8915
Validation - Loss: 0.8945, Accuracy: 0.6930, F1: 0.6930, AUC: 0.6930, Precision: 0.6931,  Recall: 0.6930


Epoch 8/30: 100%|██████████| 13137/13137 [00:44<00:00, 292.62it/s]


Epoch 8/30, Average Train Loss: 0.8948
Validation - Loss: 0.8947, Accuracy: 0.6961, F1: 0.6957, AUC: 0.6960, Precision: 0.6972,  Recall: 0.6961


Epoch 9/30: 100%|██████████| 13137/13137 [00:45<00:00, 286.15it/s]


Epoch 9/30, Average Train Loss: 0.8973
Validation - Loss: 0.8976, Accuracy: 0.6960, F1: 0.6957, AUC: 0.6959, Precision: 0.6968,  Recall: 0.6960
Learning rate reduced to 0.0025


Epoch 10/30: 100%|██████████| 13137/13137 [00:46<00:00, 284.88it/s]


Epoch 10/30, Average Train Loss: 0.8992
Validation - Loss: 0.8983, Accuracy: 0.6961, F1: 0.6957, AUC: 0.6960, Precision: 0.6970,  Recall: 0.6961


Epoch 11/30: 100%|██████████| 13137/13137 [00:45<00:00, 285.66it/s]


Epoch 11/30, Average Train Loss: 0.9005
Validation - Loss: 0.9001, Accuracy: 0.6967, F1: 0.6958, AUC: 0.6965, Precision: 0.6988,  Recall: 0.6967


Epoch 12/30: 100%|██████████| 13137/13137 [00:45<00:00, 290.55it/s]


Epoch 12/30, Average Train Loss: 0.9021
Validation - Loss: 0.9023, Accuracy: 0.6967, F1: 0.6957, AUC: 0.6966, Precision: 0.6993,  Recall: 0.6967


Epoch 13/30: 100%|██████████| 13137/13137 [00:47<00:00, 279.00it/s]


Epoch 13/30, Average Train Loss: 0.9038
Validation - Loss: 0.9039, Accuracy: 0.6967, F1: 0.6952, AUC: 0.6965, Precision: 0.7004,  Recall: 0.6967
Learning rate reduced to 0.00125


Epoch 14/30: 100%|██████████| 13137/13137 [00:45<00:00, 288.53it/s]


Epoch 14/30, Average Train Loss: 0.9047
Validation - Loss: 0.9038, Accuracy: 0.6970, F1: 0.6963, AUC: 0.6969, Precision: 0.6987,  Recall: 0.6970


Epoch 15/30: 100%|██████████| 13137/13137 [00:45<00:00, 287.64it/s]


Epoch 15/30, Average Train Loss: 0.9053
Validation - Loss: 0.9053, Accuracy: 0.6970, F1: 0.6963, AUC: 0.6969, Precision: 0.6989,  Recall: 0.6970


Epoch 16/30: 100%|██████████| 13137/13137 [00:46<00:00, 282.42it/s]


Epoch 16/30, Average Train Loss: 0.9064
Validation - Loss: 0.9051, Accuracy: 0.6973, F1: 0.6970, AUC: 0.6972, Precision: 0.6980,  Recall: 0.6973


Epoch 17/30: 100%|██████████| 13137/13137 [00:45<00:00, 291.43it/s]


Epoch 17/30, Average Train Loss: 0.9069
Validation - Loss: 0.9062, Accuracy: 0.6975, F1: 0.6966, AUC: 0.6974, Precision: 0.6997,  Recall: 0.6975
Learning rate reduced to 0.000625


Epoch 18/30: 100%|██████████| 13137/13137 [00:45<00:00, 285.59it/s]


Epoch 18/30, Average Train Loss: 0.9074
Validation - Loss: 0.9070, Accuracy: 0.6971, F1: 0.6965, AUC: 0.6970, Precision: 0.6983,  Recall: 0.6971


Epoch 19/30: 100%|██████████| 13137/13137 [00:44<00:00, 291.96it/s]


Epoch 19/30, Average Train Loss: 0.9079
Validation - Loss: 0.9068, Accuracy: 0.6978, F1: 0.6971, AUC: 0.6977, Precision: 0.6996,  Recall: 0.6978


Epoch 20/30: 100%|██████████| 13137/13137 [00:45<00:00, 287.45it/s]


Epoch 20/30, Average Train Loss: 0.9083
Validation - Loss: 0.9079, Accuracy: 0.6972, F1: 0.6964, AUC: 0.6971, Precision: 0.6992,  Recall: 0.6972


Epoch 21/30: 100%|██████████| 13137/13137 [00:44<00:00, 294.22it/s]


Epoch 21/30, Average Train Loss: 0.9086
Validation - Loss: 0.9078, Accuracy: 0.6974, F1: 0.6967, AUC: 0.6972, Precision: 0.6990,  Recall: 0.6974
Learning rate reduced to 0.0003125


Epoch 22/30: 100%|██████████| 13137/13137 [00:44<00:00, 292.07it/s]


Epoch 22/30, Average Train Loss: 0.9089
Validation - Loss: 0.9084, Accuracy: 0.6977, F1: 0.6967, AUC: 0.6976, Precision: 0.7003,  Recall: 0.6977


Epoch 23/30: 100%|██████████| 13137/13137 [00:46<00:00, 282.70it/s]


Epoch 23/30, Average Train Loss: 0.9091
Validation - Loss: 0.9083, Accuracy: 0.6979, F1: 0.6970, AUC: 0.6977, Precision: 0.6999,  Recall: 0.6979


Epoch 24/30: 100%|██████████| 13137/13137 [00:45<00:00, 291.80it/s]


Epoch 24/30, Average Train Loss: 0.9093
Validation - Loss: 0.9088, Accuracy: 0.6977, F1: 0.6970, AUC: 0.6976, Precision: 0.6993,  Recall: 0.6977


Epoch 25/30: 100%|██████████| 13137/13137 [00:45<00:00, 287.11it/s]


Epoch 25/30, Average Train Loss: 0.9096
Validation - Loss: 0.9089, Accuracy: 0.6977, F1: 0.6966, AUC: 0.6975, Precision: 0.7003,  Recall: 0.6977
Learning rate reduced to 0.00015625


Epoch 26/30: 100%|██████████| 13137/13137 [00:45<00:00, 287.51it/s]


Epoch 26/30, Average Train Loss: 0.9097
Validation - Loss: 0.9089, Accuracy: 0.6978, F1: 0.6970, AUC: 0.6976, Precision: 0.6997,  Recall: 0.6978


Epoch 27/30: 100%|██████████| 13137/13137 [00:45<00:00, 288.22it/s]


Epoch 27/30, Average Train Loss: 0.9098
Validation - Loss: 0.9089, Accuracy: 0.6979, F1: 0.6971, AUC: 0.6977, Precision: 0.6999,  Recall: 0.6979


Epoch 28/30: 100%|██████████| 13137/13137 [00:44<00:00, 293.74it/s]


Epoch 28/30, Average Train Loss: 0.9100
Validation - Loss: 0.9093, Accuracy: 0.6976, F1: 0.6965, AUC: 0.6974, Precision: 0.7001,  Recall: 0.6976


Epoch 29/30: 100%|██████████| 13137/13137 [00:46<00:00, 284.34it/s]


Epoch 29/30, Average Train Loss: 0.9100
Validation - Loss: 0.9092, Accuracy: 0.6981, F1: 0.6974, AUC: 0.6979, Precision: 0.6997,  Recall: 0.6981
Learning rate reduced to 7.8125e-05


Epoch 30/30: 100%|██████████| 13137/13137 [00:45<00:00, 291.20it/s]


Epoch 30/30, Average Train Loss: 0.9102
Validation - Loss: 0.9092, Accuracy: 0.6980, F1: 0.6973, AUC: 0.6978, Precision: 0.6996,  Recall: 0.6980
Final Validation - Accuracy: 0.6980, F1: 0.6973, AUC: 0.6978

Fold 2/3


Epoch 1/30: 100%|██████████| 13137/13137 [00:45<00:00, 286.77it/s]


Epoch 1/30, Average Train Loss: 0.8752
Validation - Loss: 0.8725, Accuracy: 0.6921, F1: 0.6916, AUC: 0.6921, Precision: 0.6933,  Recall: 0.6921
New best model saved


Epoch 2/30: 100%|██████████| 13137/13137 [00:45<00:00, 286.50it/s]


Epoch 2/30, Average Train Loss: 0.8753
Validation - Loss: 0.8815, Accuracy: 0.6892, F1: 0.6859, AUC: 0.6893, Precision: 0.6977,  Recall: 0.6892


Epoch 3/30: 100%|██████████| 13137/13137 [00:45<00:00, 288.20it/s]


Epoch 3/30, Average Train Loss: 0.8780
Validation - Loss: 0.8802, Accuracy: 0.6902, F1: 0.6872, AUC: 0.6903, Precision: 0.6979,  Recall: 0.6902


Epoch 4/30: 100%|██████████| 13137/13137 [00:44<00:00, 292.22it/s]


Epoch 4/30, Average Train Loss: 0.8822
Validation - Loss: 0.8826, Accuracy: 0.6932, F1: 0.6926, AUC: 0.6932, Precision: 0.6945,  Recall: 0.6932


Epoch 5/30: 100%|██████████| 13137/13137 [00:45<00:00, 286.43it/s]


Epoch 5/30, Average Train Loss: 0.8875
Validation - Loss: 0.8887, Accuracy: 0.6923, F1: 0.6896, AUC: 0.6924, Precision: 0.6994,  Recall: 0.6923
Learning rate reduced to 0.005


Epoch 6/30: 100%|██████████| 13137/13137 [00:45<00:00, 287.49it/s]


Epoch 6/30, Average Train Loss: 0.8893
Validation - Loss: 0.8880, Accuracy: 0.6955, F1: 0.6951, AUC: 0.6955, Precision: 0.6964,  Recall: 0.6955


Epoch 7/30: 100%|██████████| 13137/13137 [00:44<00:00, 292.20it/s]


Epoch 7/30, Average Train Loss: 0.8917
Validation - Loss: 0.8931, Accuracy: 0.6945, F1: 0.6930, AUC: 0.6945, Precision: 0.6986,  Recall: 0.6945


Epoch 8/30: 100%|██████████| 13137/13137 [00:45<00:00, 287.59it/s]


Epoch 8/30, Average Train Loss: 0.8945
Validation - Loss: 0.8957, Accuracy: 0.6950, F1: 0.6949, AUC: 0.6950, Precision: 0.6951,  Recall: 0.6950


Epoch 9/30: 100%|██████████| 13137/13137 [00:44<00:00, 291.94it/s]


Epoch 9/30, Average Train Loss: 0.8975
Validation - Loss: 0.8968, Accuracy: 0.6959, F1: 0.6955, AUC: 0.6959, Precision: 0.6970,  Recall: 0.6959
Learning rate reduced to 0.0025


Epoch 10/30: 100%|██████████| 13137/13137 [00:45<00:00, 288.17it/s]


Epoch 10/30, Average Train Loss: 0.8983
Validation - Loss: 0.8983, Accuracy: 0.6977, F1: 0.6976, AUC: 0.6978, Precision: 0.6981,  Recall: 0.6977


Epoch 11/30: 100%|██████████| 13137/13137 [00:45<00:00, 289.80it/s]


Epoch 11/30, Average Train Loss: 0.8997
Validation - Loss: 0.8994, Accuracy: 0.6957, F1: 0.6948, AUC: 0.6958, Precision: 0.6981,  Recall: 0.6957


Epoch 12/30: 100%|██████████| 13137/13137 [00:45<00:00, 289.31it/s]


Epoch 12/30, Average Train Loss: 0.9016
Validation - Loss: 0.9015, Accuracy: 0.6966, F1: 0.6957, AUC: 0.6966, Precision: 0.6987,  Recall: 0.6966


Epoch 13/30: 100%|██████████| 13137/13137 [00:45<00:00, 287.87it/s]


Epoch 13/30, Average Train Loss: 0.9033
Validation - Loss: 0.9019, Accuracy: 0.6971, F1: 0.6970, AUC: 0.6972, Precision: 0.6975,  Recall: 0.6971
Learning rate reduced to 0.00125


Epoch 14/30: 100%|██████████| 13137/13137 [00:44<00:00, 292.07it/s]


Epoch 14/30, Average Train Loss: 0.9043
Validation - Loss: 0.9032, Accuracy: 0.6965, F1: 0.6954, AUC: 0.6966, Precision: 0.6996,  Recall: 0.6965


Epoch 15/30: 100%|██████████| 13137/13137 [00:46<00:00, 280.08it/s]


Epoch 15/30, Average Train Loss: 0.9049
Validation - Loss: 0.9045, Accuracy: 0.6969, F1: 0.6954, AUC: 0.6970, Precision: 0.7009,  Recall: 0.6969


Epoch 16/30: 100%|██████████| 13137/13137 [00:45<00:00, 291.72it/s]


Epoch 16/30, Average Train Loss: 0.9060
Validation - Loss: 0.9046, Accuracy: 0.6966, F1: 0.6959, AUC: 0.6967, Precision: 0.6985,  Recall: 0.6966


Epoch 17/30: 100%|██████████| 13137/13137 [00:45<00:00, 287.78it/s]


Epoch 17/30, Average Train Loss: 0.9065
Validation - Loss: 0.9056, Accuracy: 0.6975, F1: 0.6971, AUC: 0.6975, Precision: 0.6986,  Recall: 0.6975
Learning rate reduced to 0.000625


Epoch 18/30: 100%|██████████| 13137/13137 [00:46<00:00, 285.58it/s]


Epoch 18/30, Average Train Loss: 0.9072
Validation - Loss: 0.9060, Accuracy: 0.6978, F1: 0.6974, AUC: 0.6978, Precision: 0.6988,  Recall: 0.6978


Epoch 19/30: 100%|██████████| 13137/13137 [00:45<00:00, 290.87it/s]


Epoch 19/30, Average Train Loss: 0.9078
Validation - Loss: 0.9060, Accuracy: 0.6973, F1: 0.6969, AUC: 0.6973, Precision: 0.6985,  Recall: 0.6973


Epoch 20/30: 100%|██████████| 13137/13137 [00:45<00:00, 288.45it/s]


Epoch 20/30, Average Train Loss: 0.9080
Validation - Loss: 0.9069, Accuracy: 0.6973, F1: 0.6968, AUC: 0.6974, Precision: 0.6989,  Recall: 0.6973


Epoch 21/30: 100%|██████████| 13137/13137 [00:45<00:00, 286.19it/s]


Epoch 21/30, Average Train Loss: 0.9085
Validation - Loss: 0.9072, Accuracy: 0.6972, F1: 0.6965, AUC: 0.6972, Precision: 0.6989,  Recall: 0.6972
Learning rate reduced to 0.0003125


Epoch 22/30: 100%|██████████| 13137/13137 [00:46<00:00, 281.79it/s]


Epoch 22/30, Average Train Loss: 0.9088
Validation - Loss: 0.9077, Accuracy: 0.6970, F1: 0.6962, AUC: 0.6971, Precision: 0.6993,  Recall: 0.6970


Epoch 23/30: 100%|██████████| 13137/13137 [00:47<00:00, 274.17it/s]


Epoch 23/30, Average Train Loss: 0.9090
Validation - Loss: 0.9076, Accuracy: 0.6975, F1: 0.6970, AUC: 0.6975, Precision: 0.6987,  Recall: 0.6975


Epoch 24/30: 100%|██████████| 13137/13137 [00:46<00:00, 284.72it/s]


Epoch 24/30, Average Train Loss: 0.9093
Validation - Loss: 0.9077, Accuracy: 0.6973, F1: 0.6966, AUC: 0.6973, Precision: 0.6989,  Recall: 0.6973


Epoch 25/30: 100%|██████████| 13137/13137 [00:46<00:00, 284.45it/s]


Epoch 25/30, Average Train Loss: 0.9095
Validation - Loss: 0.9084, Accuracy: 0.6972, F1: 0.6963, AUC: 0.6972, Precision: 0.6995,  Recall: 0.6972
Learning rate reduced to 0.00015625


Epoch 26/30: 100%|██████████| 13137/13137 [00:45<00:00, 289.65it/s]


Epoch 26/30, Average Train Loss: 0.9097
Validation - Loss: 0.9082, Accuracy: 0.6974, F1: 0.6967, AUC: 0.6974, Precision: 0.6993,  Recall: 0.6974


Epoch 27/30: 100%|██████████| 13137/13137 [00:47<00:00, 279.06it/s]


Epoch 27/30, Average Train Loss: 0.9095
Validation - Loss: 0.9082, Accuracy: 0.6975, F1: 0.6968, AUC: 0.6975, Precision: 0.6992,  Recall: 0.6975


Epoch 28/30: 100%|██████████| 13137/13137 [00:45<00:00, 288.53it/s]


Epoch 28/30, Average Train Loss: 0.9099
Validation - Loss: 0.9085, Accuracy: 0.6973, F1: 0.6965, AUC: 0.6973, Precision: 0.6994,  Recall: 0.6973


Epoch 29/30: 100%|██████████| 13137/13137 [00:45<00:00, 289.27it/s]


Epoch 29/30, Average Train Loss: 0.9099
Validation - Loss: 0.9084, Accuracy: 0.6974, F1: 0.6968, AUC: 0.6975, Precision: 0.6991,  Recall: 0.6974
Learning rate reduced to 7.8125e-05


Epoch 30/30: 100%|██████████| 13137/13137 [00:47<00:00, 277.03it/s]


Epoch 30/30, Average Train Loss: 0.9099
Validation - Loss: 0.9085, Accuracy: 0.6974, F1: 0.6968, AUC: 0.6975, Precision: 0.6992,  Recall: 0.6974
Final Validation - Accuracy: 0.6974, F1: 0.6968, AUC: 0.6975

Fold 3/3


Epoch 1/30: 100%|██████████| 13137/13137 [00:45<00:00, 287.26it/s]


Epoch 1/30, Average Train Loss: 0.8760
Validation - Loss: 0.8719, Accuracy: 0.6924, F1: 0.6917, AUC: 0.6925, Precision: 0.6945,  Recall: 0.6924
New best model saved


Epoch 2/30: 100%|██████████| 13137/13137 [00:45<00:00, 290.75it/s]


Epoch 2/30, Average Train Loss: 0.8767
Validation - Loss: 0.8765, Accuracy: 0.6909, F1: 0.6902, AUC: 0.6910, Precision: 0.6928,  Recall: 0.6909


Epoch 3/30: 100%|██████████| 13137/13137 [00:46<00:00, 285.57it/s]


Epoch 3/30, Average Train Loss: 0.8800
Validation - Loss: 0.8811, Accuracy: 0.6902, F1: 0.6881, AUC: 0.6903, Precision: 0.6956,  Recall: 0.6902


Epoch 4/30: 100%|██████████| 13137/13137 [00:45<00:00, 288.18it/s]


Epoch 4/30, Average Train Loss: 0.8836
Validation - Loss: 0.8831, Accuracy: 0.6931, F1: 0.6931, AUC: 0.6931, Precision: 0.6931,  Recall: 0.6931


Epoch 5/30: 100%|██████████| 13137/13137 [00:44<00:00, 292.84it/s]


Epoch 5/30, Average Train Loss: 0.8884
Validation - Loss: 0.8912, Accuracy: 0.6925, F1: 0.6904, AUC: 0.6926, Precision: 0.6979,  Recall: 0.6925
Learning rate reduced to 0.005


Epoch 6/30: 100%|██████████| 13137/13137 [00:46<00:00, 281.07it/s]


Epoch 6/30, Average Train Loss: 0.8910
Validation - Loss: 0.8902, Accuracy: 0.6945, F1: 0.6942, AUC: 0.6946, Precision: 0.6954,  Recall: 0.6945


Epoch 7/30: 100%|██████████| 13137/13137 [00:44<00:00, 293.99it/s]


Epoch 7/30, Average Train Loss: 0.8935
Validation - Loss: 0.8935, Accuracy: 0.6940, F1: 0.6935, AUC: 0.6941, Precision: 0.6955,  Recall: 0.6940


Epoch 8/30: 100%|██████████| 13137/13137 [00:44<00:00, 294.27it/s]


Epoch 8/30, Average Train Loss: 0.8966
Validation - Loss: 0.8952, Accuracy: 0.6964, F1: 0.6957, AUC: 0.6965, Precision: 0.6982,  Recall: 0.6964


Epoch 9/30: 100%|██████████| 13137/13137 [00:46<00:00, 284.25it/s]


Epoch 9/30, Average Train Loss: 0.8990
Validation - Loss: 0.8980, Accuracy: 0.6955, F1: 0.6945, AUC: 0.6956, Precision: 0.6984,  Recall: 0.6955
Learning rate reduced to 0.0025


Epoch 10/30: 100%|██████████| 13137/13137 [00:45<00:00, 291.46it/s]


Epoch 10/30, Average Train Loss: 0.9007
Validation - Loss: 0.8983, Accuracy: 0.6971, F1: 0.6968, AUC: 0.6972, Precision: 0.6982,  Recall: 0.6971


Epoch 11/30: 100%|██████████| 13137/13137 [00:45<00:00, 287.40it/s]


Epoch 11/30, Average Train Loss: 0.9017
Validation - Loss: 0.8998, Accuracy: 0.6964, F1: 0.6961, AUC: 0.6964, Precision: 0.6971,  Recall: 0.6964


Epoch 12/30: 100%|██████████| 13137/13137 [00:45<00:00, 286.55it/s]


Epoch 12/30, Average Train Loss: 0.9036
Validation - Loss: 0.9020, Accuracy: 0.6962, F1: 0.6955, AUC: 0.6963, Precision: 0.6983,  Recall: 0.6962


Epoch 13/30: 100%|██████████| 13137/13137 [00:45<00:00, 288.15it/s]


Epoch 13/30, Average Train Loss: 0.9050
Validation - Loss: 0.9039, Accuracy: 0.6965, F1: 0.6959, AUC: 0.6966, Precision: 0.6982,  Recall: 0.6965
Learning rate reduced to 0.00125


Epoch 14/30: 100%|██████████| 13137/13137 [00:44<00:00, 296.00it/s]


Epoch 14/30, Average Train Loss: 0.9056
Validation - Loss: 0.9037, Accuracy: 0.6967, F1: 0.6959, AUC: 0.6968, Precision: 0.6991,  Recall: 0.6967


Epoch 15/30: 100%|██████████| 13137/13137 [00:46<00:00, 283.94it/s]


Epoch 15/30, Average Train Loss: 0.9064
Validation - Loss: 0.9047, Accuracy: 0.6967, F1: 0.6959, AUC: 0.6968, Precision: 0.6990,  Recall: 0.6967


Epoch 16/30: 100%|██████████| 13137/13137 [00:44<00:00, 293.02it/s]


Epoch 16/30, Average Train Loss: 0.9073
Validation - Loss: 0.9049, Accuracy: 0.6974, F1: 0.6971, AUC: 0.6975, Precision: 0.6983,  Recall: 0.6974


Epoch 17/30: 100%|██████████| 13137/13137 [00:45<00:00, 288.87it/s]


Epoch 17/30, Average Train Loss: 0.9079
Validation - Loss: 0.9063, Accuracy: 0.6964, F1: 0.6954, AUC: 0.6965, Precision: 0.6992,  Recall: 0.6964
Learning rate reduced to 0.000625


Epoch 18/30: 100%|██████████| 13137/13137 [00:45<00:00, 287.71it/s]


Epoch 18/30, Average Train Loss: 0.9085
Validation - Loss: 0.9059, Accuracy: 0.6977, F1: 0.6973, AUC: 0.6978, Precision: 0.6990,  Recall: 0.6977


Epoch 19/30: 100%|██████████| 13137/13137 [00:44<00:00, 292.68it/s]


Epoch 19/30, Average Train Loss: 0.9085
Validation - Loss: 0.9067, Accuracy: 0.6972, F1: 0.6962, AUC: 0.6973, Precision: 0.6999,  Recall: 0.6972


Epoch 20/30: 100%|██████████| 13137/13137 [00:45<00:00, 290.85it/s]


Epoch 20/30, Average Train Loss: 0.9091
Validation - Loss: 0.9070, Accuracy: 0.6971, F1: 0.6964, AUC: 0.6972, Precision: 0.6991,  Recall: 0.6971


Epoch 21/30: 100%|██████████| 13137/13137 [00:45<00:00, 286.30it/s]


Epoch 21/30, Average Train Loss: 0.9095
Validation - Loss: 0.9077, Accuracy: 0.6972, F1: 0.6962, AUC: 0.6973, Precision: 0.7000,  Recall: 0.6972
Learning rate reduced to 0.0003125


Epoch 22/30: 100%|██████████| 13137/13137 [00:45<00:00, 290.72it/s]


Epoch 22/30, Average Train Loss: 0.9095
Validation - Loss: 0.9076, Accuracy: 0.6977, F1: 0.6972, AUC: 0.6978, Precision: 0.6992,  Recall: 0.6977


Epoch 23/30: 100%|██████████| 13137/13137 [00:44<00:00, 296.21it/s]


Epoch 23/30, Average Train Loss: 0.9096
Validation - Loss: 0.9080, Accuracy: 0.6976, F1: 0.6968, AUC: 0.6976, Precision: 0.6996,  Recall: 0.6976


Epoch 24/30: 100%|██████████| 13137/13137 [00:46<00:00, 284.65it/s]


Epoch 24/30, Average Train Loss: 0.9100
Validation - Loss: 0.9078, Accuracy: 0.6974, F1: 0.6966, AUC: 0.6975, Precision: 0.6994,  Recall: 0.6974


Epoch 25/30: 100%|██████████| 13137/13137 [00:45<00:00, 290.63it/s]


Epoch 25/30, Average Train Loss: 0.9104
Validation - Loss: 0.9086, Accuracy: 0.6970, F1: 0.6961, AUC: 0.6971, Precision: 0.6997,  Recall: 0.6970
Learning rate reduced to 0.00015625


Epoch 26/30: 100%|██████████| 13137/13137 [00:44<00:00, 295.62it/s]


Epoch 26/30, Average Train Loss: 0.9103
Validation - Loss: 0.9083, Accuracy: 0.6977, F1: 0.6971, AUC: 0.6978, Precision: 0.6995,  Recall: 0.6977


Epoch 27/30: 100%|██████████| 13137/13137 [00:46<00:00, 283.60it/s]


Epoch 27/30, Average Train Loss: 0.9105
Validation - Loss: 0.9083, Accuracy: 0.6977, F1: 0.6972, AUC: 0.6978, Precision: 0.6991,  Recall: 0.6977


Epoch 28/30: 100%|██████████| 13137/13137 [00:44<00:00, 294.68it/s]


Epoch 28/30, Average Train Loss: 0.9105
Validation - Loss: 0.9085, Accuracy: 0.6974, F1: 0.6966, AUC: 0.6974, Precision: 0.6994,  Recall: 0.6974


Epoch 29/30: 100%|██████████| 13137/13137 [00:45<00:00, 291.49it/s]


Epoch 29/30, Average Train Loss: 0.9106
Validation - Loss: 0.9085, Accuracy: 0.6976, F1: 0.6970, AUC: 0.6977, Precision: 0.6994,  Recall: 0.6976
Learning rate reduced to 7.8125e-05


Epoch 30/30: 100%|██████████| 13137/13137 [00:45<00:00, 285.77it/s]


Epoch 30/30, Average Train Loss: 0.9107
Validation - Loss: 0.9088, Accuracy: 0.6973, F1: 0.6965, AUC: 0.6974, Precision: 0.6994,  Recall: 0.6973
Final Validation - Accuracy: 0.6973, F1: 0.6965, AUC: 0.6974
Training completed. Best model loaded.
Final Validation - Accuracy: 0.6973, F1: 0.6965, AUC: 0.6974


In [10]:
print("\nCross-validation results:")
print(f"Mean Accuracy: {np.mean(cv_accuracies):.4f} (+/- {np.std(cv_accuracies):.4f})")
print(f"Mean F1 Score: {np.mean(cv_f1_scores):.4f} (+/- {np.std(cv_f1_scores):.4f})")
print(f"Mean AUC Score: {np.mean(val_auc):.4f} (+/- {np.std(val_auc):.4f})")
print(f"Mean Precision: {np.mean(val_precision):.4f} (+/- {np.std(val_precision):.4f})")
print(f"Mean Recall: {np.mean(val_recall):.4f} (+/- {np.std(val_recall):.4f})")


Cross-validation results:
Mean Accuracy: 0.6975 (+/- 0.0003)
Mean F1 Score: 0.6968 (+/- 0.0003)
Mean AUC Score: 0.6974 (+/- 0.0000)
Mean Precision: 0.6994 (+/- 0.0000)
Mean Recall: 0.6973 (+/- 0.0000)


Cross-validation results:
- Mean Accuracy: 0.6975 (+/- 0.0003)
- Mean F1 Score: 0.6968 (+/- 0.0003)
- Mean AUC Score: 0.6974 (+/- 0.0000)
- Mean Precision: 0.6994 (+/- 0.0000)
- Mean Recall: 0.6973 (+/- 0.0000)

## Using XGBoost

In [11]:
# Split the resampled training data into training and validation sets
X_train_final, X_val, y_train_final, y_val = train_test_split(X_train_resampled, y_train_resampled, test_size=0.2, random_state=42)

# Define XGBoost model
xgb_model = xgb.XGBClassifier(use_label_encoder=False, eval_metric='logloss', random_state=42)

# Define parameter grid for GridSearchCV
param_grid = {
    'max_depth': [5, 7],
    'learning_rate': [0.1, 0.3],
    'n_estimators': [200, 300],
    'min_child_weight': [1, 3]
}

# Perform GridSearchCV
grid_search = GridSearchCV(estimator=xgb_model, param_grid=param_grid,
                           cv=3, scoring='roc_auc', verbose=2, n_jobs=-1)
grid_search.fit(X_train_final, y_train_final)

# Get the best model
best_xgb_model = grid_search.best_estimator_

# Train the best model on the entire training set
best_xgb_model.fit(X_train_resampled, y_train_resampled)

# Make predictions on the OOT test set
y_oot_pred = best_xgb_model.predict(X_oot_test_scaled)
y_oot_pred_proba = best_xgb_model.predict_proba(X_oot_test_scaled)[:, 1]

# Calculate metrics
accuracy = accuracy_score(y_oot_test, y_oot_pred)
precision, recall, f1, _ = precision_recall_fscore_support(y_oot_test, y_oot_pred, average='weighted', zero_division=0)
auc = roc_auc_score(y_oot_test, y_oot_pred_proba)

# Print results
print("\nXGBoost Final OOT Test Results:")
print(f"Accuracy: {accuracy:.4f}")
print(f"Precision: {precision:.4f}")
print(f"Recall: {recall:.4f}")
print(f"F1 Score: {f1:.4f}")
print(f"AUC: {auc:.4f}")

# Print best parameters
print("\nBest XGBoost Parameters:")
print(grid_search.best_params_)

# Feature importance
feature_importance = best_xgb_model.feature_importances_
sorted_idx = np.argsort(feature_importance)
pos = np.arange(sorted_idx.shape[0]) + .5

Fitting 3 folds for each of 16 candidates, totalling 48 fits


/usr/local/lib/python3.10/dist-packages/xgboost/core.py:158: UserWarning: [19:09:30] WARNING: /workspace/src/learner.cc:740: 
Parameters: { "use_label_encoder" } are not used.

  warnings.warn(smsg, UserWarning)
/usr/local/lib/python3.10/dist-packages/xgboost/core.py:158: UserWarning: [19:09:39] WARNING: /workspace/src/learner.cc:740: 
Parameters: { "use_label_encoder" } are not used.

  warnings.warn(smsg, UserWarning)



XGBoost Final OOT Test Results:
Accuracy: 0.9310
Precision: 0.9796
Recall: 0.9310
F1 Score: 0.9542
AUC: 0.6666

Best XGBoost Parameters:
{'learning_rate': 0.3, 'max_depth': 7, 'min_child_weight': 1, 'n_estimators': 300}

Feature Importances:
tensor([-0.0092, -0.2972, -0.0982,  0.1020, -0.7515,  0.1949, -0.8050],
       device='cuda:0'): 0.0330
tensor([-0.6412, -0.2972, -0.0982, -0.3785, -0.1352, -0.2842, -0.8050],
       device='cuda:0'): 0.0496


IndexError: index 3 is out of bounds for dimension 0 with size 3

XGBoost Final OOT Test Results:
- Accuracy: 0.9310
- Precision: 0.9796
- Recall: 0.9310
-F1 Score: 0.9542
- AUC: 0.6666
- Best XGBoost Parameters: {'learning_rate': 0.3, 'max_depth': 7, 'min_child_weight': 1, 'n_estimators': 300}".